File: waltz-model/src/main/java/com/khartec/waltz/model/appview/AppView.java
/*
 *  This file is part of Waltz.
 *
 *     Waltz is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     Waltz is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with Waltz.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.khartec.waltz.model.appview;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.khartec.waltz.model.application.Application;
import com.khartec.waltz.model.applicationcapability.ApplicationCapability;
import com.khartec.waltz.model.bookmark.Bookmark;
import com.khartec.waltz.model.capability.Capability;
import com.khartec.waltz.model.cost.AssetCost;
import com.khartec.waltz.model.orgunit.OrganisationalUnit;
import com.khartec.waltz.model.trait.Trait;
import org.immutables.value.Value;

import java.util.Collection;
import java.util.List;


@Value.Immutable
@JsonSerialize(as = ImmutableAppView.class)
@JsonDeserialize(as = ImmutableAppView.class)
public abstract class AppView {

    public abstract Application app();
    public abstract List<String> tags();
    public abstract List<String> aliases();
    public abstract OrganisationalUnit organisationalUnit();
    public abstract Collection<Bookmark> bookmarks();
    public abstract Collection<ApplicationCapability> appCapabilities();
    public abstract Collection<Capability> capabilities();
    public abstract List<AssetCost> costs();
    public abstract List<Trait> explicitTraits();
}

File: waltz-ng/client/applications/app-view.js
/*
 * Waltz - Enterprise Architecture
 * Copyright (C) 2016  Khartec Ltd.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { CORE_API } from "../common/services/core-api-utils";
import { CORE_API } from "../common/services/core-api-utils";

import template from "./app-view.html";


const initialState = {
    app: {}
};


const addToHistory = (historyStore, app) => {
    if (! app) { return; }
    historyStore.put(
        app.name,
        'APPLICATION',
        'main.app.view',
        { id: app.id });
};


function controller($stateParams,
                    serviceBroker,
                    historyStore) {

    const id = $stateParams.id;
    const entityReference = { id, kind: 'APPLICATION' };
    const vm = Object.assign(this, initialState);

    vm.entityRef = entityReference;


    function loadAll() {
        serviceBroker
            .loadViewData(CORE_API.ApplicationStore.getById, [id])
            .then(r => vm.app = r.data)
            .then(() => postLoadActions());
    }


    function postLoadActions() {
        addToHistory(historyStore, vm.app);
        vm.entityRef = Object.assign({}, vm.entityRef, {name: vm.app.name});
    }

    // load everything in priority order
    loadAll();
}


controller.$inject = [
    '$stateParams',
    'ServiceBroker',
    'HistoryStore'
];


export default  {
    template,
    controller,
    controllerAs: 'ctrl'
};

